src/file_chooser_dialog.rs

```diff
@@ -5,21 +5,85 @@
 use ffi;
 use glib::translate::*;
 use glib::object::{Downcast, IsA};
+use libc::c_char;
 use std::ptr;
 use FileChooserAction;
 use FileChooserDialog;
+use ResponseType;
 use Widget;
 use Window;
 
 impl FileChooserDialog {
-    pub fn new<T: IsA<Window>>(title: Option<&str>, parent: Option<&T>, action: FileChooserAction)
-            -> FileChooserDialog {
+    // TODO: Keep the other constructor with buttons support as the only constructor (this one was
+    //       left for compatibility) and rename it to `new` for consistency.
+    pub fn new<T: IsA<Window>>(title: Option<&str>, parent: Option<&T>, action: FileChooserAction) -> FileChooserDialog {
         assert_initialized_main_thread!();
         unsafe {
-            Widget::from_glib_none(
-                ffi::gtk_file_chooser_dialog_new(title.to_glib_none().0, parent.to_glib_none().0,
-                    action.to_glib(), ptr::null_mut()))
+            Widget::from_glib_none(ffi::gtk_file_chooser_dialog_new(
+                title.to_glib_none().0,
+                parent.to_glib_none().0,
+                action.to_glib(),
+                ptr::null::<c_char>()
+            ))
                 .downcast_unchecked()
         }
     }
+
+    pub fn with_buttons<T: IsA<Window>>(title: Option<&str>, parent: Option<&T>, action: FileChooserAction, buttons: &[(&str, ResponseType)]) -> FileChooserDialog {
+        assert_initialized_main_thread!();
+        unsafe {
+            Widget::from_glib_none(match buttons.len() {
+                0 => {
+                    ffi::gtk_file_chooser_dialog_new(
+                        title.to_glib_none().0,
+                        parent.to_glib_none().0,
+                        action.to_glib(),
+                        ptr::null::<c_char>()
+                    )
+                },
+                1 => {
+                    ffi::gtk_file_chooser_dialog_new(
+                        title.to_glib_none().0,
+                        parent.to_glib_none().0,
+                        action.to_glib(),
+                        buttons[0].0.to_glib_none().0,
+                        buttons[0].1.to_glib(),
+                        ptr::null::<c_char>(),
+                    )
+                },
+                2 => {
+                    ffi::gtk_file_chooser_dialog_new(
+                        title.to_glib_none().0,
+                        parent.to_glib_none().0,
+                        action.to_glib(),
+                        buttons[0].0.to_glib_none().0,
+                        buttons[0].1.to_glib(),
+                        (buttons[1].0.to_glib_none() as Stash<*const c_char, str>).0,
```

Comment:
`Stash` IMHO don't needed here and in other places

Reply:
Forgot that it `var_arg` :( so rust can't infer type

Reply:
Exactly, I can omit it for the first text but not for the other ones.